---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)

library(tidyverse)
```

# annotateR

<!-- badges: start -->
<!-- badges: end -->

The goal of annotateR is to enable annotation of genome wide association study (GWAS) results (or other tabular data including chromosome, position, and allele data) with RSIDs from a reference dataset

## Installation

The development version can be installed from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("mglev1n/annotateR")
```

## Reference Datasets

```{r echo=FALSE, message=FALSE, warning=FALSE}
ref_1kg <- arrow::open_dataset("../RawData/1kg_all_phase3_snps_markers/1kg_all_phase3_snps_marker.parquet")

dplyr::tribble(
  ~dataset, ~nrow, ~ncol, ~format, ~columns,
  "1000 Genomes Phase 3", ref_1kg$num_rows, ref_1kg$num_cols, as.character(ref_1kg$format), glue::glue_collapse(ref_1kg$schema$names, sep = ", ")
) %>%
  knitr::kable()
```
